### ISSUE TITLE
Performance: watched_files_digest called many times in the context of one request

### ISSUE THREAD

**Issue by justin808** [rails/webpacker#2649]
During the scope of a single controller request, if webpacker view helpers are called multiple times, we recompute the digest of watch files multiple times. 

We could memoize this value so it runs only once in the scope of a controller request.

I'm not sure of the best way to memoize here that value.

Maybe instead, we could just put the check in `fresh?` and return true if the timestamp of the last_compilation_digest is within 2 seconds?

https://github.com/rails/webpacker/blob/master/lib/webpacker/compiler.rb#L37
```ruby
  # Returns true if all the compiled packs are up to date with the underlying asset files.
  def fresh?
    watched_files_digest == last_compilation_digest
  end
```

https://github.com/rails/webpacker/blob/master/lib/webpacker/compiler.rb#L54
```ruby
    def watched_files_digest
      warn "Webpacker::Compiler.watched_paths has been deprecated. Set additional_paths in webpacker.yml instead." unless watched_paths.empty?

      files = Dir[*default_watched_paths, *watched_paths].reject { |f| File.directory?(f) }
      file_ids = files.sort.map { |f| "#{File.basename(f)}/#{Digest::SHA1.file(f).hexdigest}" }
      Digest::SHA1.hexdigest(file_ids.join("/"))
    end
```

Another performance issue could be the desire to ensure compilation before running a test, but when running a large test suite, it seems unnecessary to recompute the digest for every request.

Maybe there might be a way to turn off compilation for CI? to override the webpacker.yml setting?

One other question is whether checking timestamps as opposed to comparing a digest could be more efficient. 

For example, if some JS file is newer than the manifest.json, then recompile. If the manifest.json is newer than all JS files, then don't recompile.

Git updates always use new timestamps when you pull the latest.

The only way timestamp checking can break is if somebody moves a file with an old timestamp. Thus, the moved file might be older than the manifest.json and thus would not trigger a recompilation.

**Comment by rossta** [comment]
@justin808 Looks like you're asking several questions here. These all seem like good questions to ask and could each better served by separate issues?

**Issue 1**

> watched_files_digest called many times in the context of one request

This is a good catch. I'm not sure what the best solution would be, but one idea is to let the view dictate that the Webpacker instance should create a new instance of the compiler which could provide memorize values for its lifespan (the request).

**Issue 2**

> Another performance issue could be the desire to ensure compilation before running a test, but when running a large test suite, it seems unnecessary to recompute the digest for every request. Maybe there might be a way to turn off compilation for CI? to override the webpacker.yml setting?

Good thoughts here. Maybe we should solve Issue 1 first, then tackle this as a separate, subsequent concern.

**Issue 3**

> One other question is whether checking timestamps as opposed to comparing a digest could be more efficient.

This is an interesting question to ask. Perhaps an example with some benchmarks in a separate issue report could help make this concern actionable.

**Comment by justin808** [comment]
> This is a good catch. I'm not sure what the best solution would be, but one idea is to let the view dictate that the Webpacker instance should create a new instance of the compiler which could provide memorize values for its lifespan (the request).

How would this play out if there are multiple requests going on? Do we need to worry about that case? Could we have a singleton object for the compiler?

I know that @gauravtiwari is working on some bigger changes. I'll wait for his feedback before submitting any code.

**Comment by justin808** [comment]
@guillaumebriday, @rossta Does this improvement still sound worthwhile? If so, I could invest some effort into improvements. Note, improvements would only be visible for projects with large numbers of files.
